File: apps/console/app/server/actions/audit.ts
"use server";
// SPDX-FileCopyrightText: 2026 Alethia Labs <legal@alethialabs.io>
// SPDX-License-Identifier: AGPL-3.0-only

import { authorize, currentActor } from "@/lib/authz/guard";
import { assertJobQuotaAllowed } from "@/lib/billing/job-quota";
import { signedJob } from "@/lib/db/signed-job";
import { withActorScope } from "@/lib/db";
import { jobs } from "@/lib/db/schema";
import { notifyScaler } from "@/lib/scaler";

/**
 * Queue an AUDIT job (elench "B" flow): the runner runs the verify engine over a
 * customer's EXISTING infrastructure — a bring-your-own OpenTofu/Terraform `show -json`
 * plan, or Kubernetes manifests — and posts the report to execution_metadata.verify_result.
 * Provisions nothing. Poll the result with the existing `getPlanResult`/`get_plan_result`
 * (which surfaces `verify_result` → the VerifyBlock).
 */
export async function queueAudit(
	input: string,
	kind: "plan" | "manifests",
	projectId?: string,
): Promise<{ jobId: string }> {
	// AUTHORIZE THE PROJECT, do not merely identify the caller (#2697).
	//
	// `currentActor()` answers "who is this", never "may they". `projectId` went straight into the
	// insert below, and RLS on `jobs` is ORG-scoped — so its WITH CHECK passes for any project in
	// the caller's org, and the job appears in that project's feed with its signed `verify_result`
	// readable by everyone who can see it.
	//
	// The caller is not only a person. `audit_infrastructure` (lib/ai/tools/scanner.ts) exposes
	// `projectId` to the MODEL as a free optional string, so the id can be chosen by prompt.
	//
	// The unattached path is legitimate — an audit of a plan that belongs to no project yet — so
	// identity alone still suffices when there is no id to authorize against.
	const actor = projectId
		? await authorize("audit", { type: "project", id: projectId })
		: await currentActor();
	const trimmed = input.trim();
	if (!trimmed) {
		throw new Error("Audit input is required (an OpenTofu plan JSON or k8s manifests).");
	}
	await assertJobQuotaAllowed(actor.orgId);

	const jobId = await withActorScope(actor, async (tx) => {
		const [job] = await tx
			.insert(jobs)
			.values(signedJob({
				user_id: actor.userId,
				...(projectId ? { project_id: projectId } : {}),
				job_type: "AUDIT",
				initiated_by: "user",
				status: "QUEUED",
				config_snapshot: { audit_kind: kind, audit_input: trimmed },
			}))
			.returning({ id: jobs.id });
		return job.id;
	});

	notifyScaler();
	return { jobId };
}

File: apps/runner/internal/agent/audit.go
// SPDX-FileCopyrightText: 2026 Alethia Labs OÜ <legal@alethialabs.io>
// SPDX-License-Identifier: AGPL-3.0-only

package agent

import (
	"context"
	"fmt"

	"github.com/alethialabs-io/alethialabs/packages/core/verify"
)

// executeAudit runs the elench verify engine over a customer's EXISTING infrastructure
// (the "B" flow): a bring-your-own OpenTofu/Terraform `show -json` plan, or Kubernetes
// manifests. It provisions nothing — it reads the input from the job config, evaluates
// it, and posts the verify.Report to execution_metadata.verify_result (surfaced by the
// console's get_plan_result tool + VerifyBlock, exactly like a PLAN job's gate).
func (w *Runner) executeAudit(ctx context.Context, job *Job, stdout, stderr *JobLogger) error {
	kind, _ := job.ConfigSnapshot["audit_kind"].(string)
	input, _ := job.ConfigSnapshot["audit_input"].(string)
	if input == "" {
		return fmt.Errorf("audit job has no input (audit_input is empty)")
	}

	fmt.Fprintf(stdout, "Auditing bring-your-own %s with elench...\n", auditKindLabel(kind))

	var report *verify.Report
	var err error
	switch kind {
	case "manifests":
		report, err = verify.EvaluateManifests([]byte(input))
	default: // "plan" (OpenTofu/Terraform show -json)
		plan, perr := verify.ParseCustomerPlan([]byte(input))
		if perr != nil {
			return fmt.Errorf("parse plan: %w", perr)
		}
		report, err = verify.Evaluate(ctx, plan)
	}
	if err != nil {
		return fmt.Errorf("audit: %w", err)
	}

	fmt.Fprintf(stdout, "Audit verdict: %s (%d pass, %d fail, %d warn, %d not-evaluable)\n",
		report.Verdict, report.Summary.Pass, report.Summary.Fail, report.Summary.Warn,
		report.Summary.NotEvaluable)

	_ = w.api.UpdateJobStatus(job.ID, "PROCESSING", "", map[string]any{"verify_result": report})
	return nil
}

func auditKindLabel(kind string) string {
	if kind == "manifests" {
		return "Kubernetes manifests"
	}
	return "OpenTofu/Terraform plan"
}
